File: ProjectEuler/Problem_4/sol.kt
/**
 * Note that Maximum product is
 *  999x999 = 998001
 * which fits the range of Int.
 */

fun largestPalindrome(): Int {

    var res = 0;

    /**
     * Since we are looking for "Largest" palindrome, we go from higher to lower
     */
    for (i in 1000 downTo 100 ) {
        for (j in 1000 downTo 100) {
            var product = i * j;
            if (isPalindrome(product)) {
               /**
                * We could have added an AND clause in above condition
                * but this way, we can break and reduse loop cycles.
                */
                if (product > res) {
                    res = product;
                } else {
                    break;
                }
            }

            /**
             * We can probably make an early break via conditions, here too,
             * to reduce the loop cycle, but logic is yet to be tested.
             */
        }
    }
    return res;
}

fun isPalindrome(num: Int): Boolean {
    val str = num.toString();
    /**
     * we check if first letter is same as last letter,
     * and meet at middle.
     */
    for (i in 0 .. (str.length / 2)){
        if (!str[i].equals(str[str.length-i-1])){

            return false
        }
    }
    return true
}

 fun main(){
    println(largestPalindrome())
 }

File: ProjectEuler/Problem_4/sol.cs
using System;

namespace ProjectEular
{
    class sol
    {
        protected bool isPalindrome(int n)
        {
            String str = n.ToString();
            for (int i = 0; i <= (str.Length / 2); i++)
            {
                if (str[i] != str[str.Length - 1 - i])
                {
                    return false;
                }
            }
            return true;
        }

        static void Main()
        {
            int res = 0;
            for (int i = 1000; i >= 100; i--)
            {
                for (int j = 1000; j >= 100; j--)
                {
                    int product = i*j;
                    if ((new sol()).isPalindrome(product)){
                        if (product > res ){
                            res = product;
                        }
                    }
                }
            }

            Console.WriteLine(res);
        }
    }
}
